### PR TITLE
ip pool option for EPG

### PR DESCRIPTION
**Pull request by rchirakk**
IP pool for endpoint group:
This feature enables users to specify an optional range of ip address
for EPG from the network subnet.
fixes https://github.com/contiv/netplugin/issues/710

### CODE REVIEW

File: netmaster/master/endpointGroup.go [hunk 1]
```diff
@@ -64,6 +67,32 @@ func CreateEndpointGroup(tenantName, networkName, groupName string) error {
...
+			return fmt.Errorf("ipv6 address range is not supported")
+		}
+
+		if err = netutils.ValidateNetworkRangeParams(ipPool, nwCfg.SubnetLen); err != nil {
+			return fmt.Errorf("invalid ip-pool %s", ipPool)
+		}
+
+		addrRangeList := strings.Split(ipPool, "-")
+		if _, err := netutils.GetIPNumber(nwCfg.SubnetIP, nwCfg.SubnetLen, 32, addrRangeList[0]); err != nil {
+			return fmt.Errorf("ip-pool %s is not part of network %s/%d", ipPool, nwCfg.SubnetIP,
```
Comment (jojimt) [inline netmaster/master/endpointGroup.go]: Please change the error message to "Bad ip-pool %s. Epg ip-pool must be a subset of corresponding network ip-pool"

File: netmaster/master/endpointGroup.go [hunk 2]
```diff
@@ -64,6 +67,32 @@ func CreateEndpointGroup(tenantName, networkName, groupName string) error {
...
+				nwCfg.SubnetLen)
+		}
+		if _, err := netutils.GetIPNumber(nwCfg.SubnetIP, nwCfg.SubnetLen, 32, addrRangeList[1]); err != nil {
+			return fmt.Errorf("ip-pool %s is not part of network %s/%d", ipPool, nwCfg.SubnetIP,
+				nwCfg.SubnetLen)
```
Comment (jojimt) [inline netmaster/master/endpointGroup.go]: Pls change per above.

File: test/integration/endpoint_test.go [hunk 3]
```diff
@@ -214,6 +218,117 @@ func (its *integTestSuite) TestEndpointGroupCreateDelete(c *C) {
...
+	// create an endpoint in the network
+	epCfg1, err := its.createEndpoint("default", "test", "epg1", addr, "")
+	assertNoErr(err, c, "creating endpoint")
+	addr, err = its.allocAddress("", "test:epg1.default", "")
+	assertErr(err, c, "allocating address")
+	err = its.deleteEndpoint("default", "test", "epg1", epCfg1)
+	assertNoErr(err, c, "deleting endpoint")
+	err = its.client.EndpointGroupDelete("default", "epg1")
+	assertNoErr(err, c, "deleting epg")
+	assertNoErr(its.client.NetworkDelete("default", "test"), c, "deleting network")
```
Comment (jojimt) [inline test/integration/endpoint_test.go]: Can we add test cases with multiple epgs on the same network, and also verify the available range and allocated range with inspect on the epg as well as the network? Can we also verify that EP gets an IP from the network pool when (1) only network is specified for the ep (2) epg is specified, but it does not have an associated ip pool.

File: netmaster/master/api.go [hunk 4]
```diff
@@ -217,17 +236,35 @@ func ReleaseAddressHandler(w http.ResponseWriter, r *http.Request, vars map[stri
...
+		s = strings.Split(s[1], ".")
+		epgName := s[0] + ":" + s[1]
+		networkID = networkID + "." + s[1]
+		epgCfg = &mastercfg.EndpointGroupState{}
+		epgCfg.StateDriver = stateDriver
+		if err = epgCfg.Read(epgName); err != nil {
+			log.Errorf("failed to read epg %s, %s", epgName, err)
+			return nil, err
+		}
+		log.Infof("AddressReleaseRequest for network: %s epg: %s", networkID, epgName)
```
Comment (rchirakk) [inline netmaster/master/api.go]: done

**Comment by rchirakk** [comment]
@jojimt  @gkvijay @abhinandanpb PTAL

**Pull request merged**